presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PickTableLayout.java

```diff
@@ -34,6 +34,10 @@
 import static com.facebook.presto.sql.planner.plan.Patterns.tableScan;
 import static java.util.Objects.requireNonNull;
 
+/**
+ * These rules should not be run after AddExchanges so as not to overwrite the TableLayout
+ * chosen by AddExchanges
```

Comment:
This could be enforced at run-time.  Just include a rule among rules returned by `PickTableLayout.rules()` that matches `Exchange` and fails.